### PR TITLE
Add new extension ``TypingChecker``

### PR DESCRIPTION
**Pull request by cdce8p**
## Steps

- [ ] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description

At the moment, the extension will provide three checkers: `deprecated-typing-alias`, `consider-using-alias`, and `consider-alternative-union-syntax`.

**Important**: For all typing related checks the checker uses the `py-version` config option instead of `sys.version_info`.

**Possible future extensions**
- [ ] Detect type comments that can be replaced by type annotations
- [ ] Detect if typing types aren't imported from `stdlib.typing` but instead another project file
- [ ] Detect imports from `typing_extensions` that are included in `typing`
- [ ] Detect the use of quotation marks in type annotations

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

--
Tests depend on https://github.com/PyCQA/astroid/pull/948 being merged and the `astroid` version being updated.

### CODE REVIEW

File: pylint/extensions/typing.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        ) and node.name in UNION_NAMES:
+            self._check_for_alternative_union_syntax(node, node.name)
+
+    @check_messages(
+        "deprecated-typing-alias",
+        "consider-using-alias",
+        "consider-alternative-union-syntax",
+    )
+    def visit_attribute(self, node: astroid.Attribute):
+        if (
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/typing.py:185]: We could create a cached function for that and the py3.10 test below it seems it's used in visit_name and visite_attribute at least.

File: doc/whatsnew/2.8.rst [hunk 2]
```diff
@@ -23,6 +23,11 @@ New checkers
 
 * Add ``consider-using-min-max-builtin`` check for if statement which could be replaced by Python builtin min or max.
 
+* Add new extension ``TypingChecker``. This optional checker can detect the use of deprecated typing aliases
+  and can suggest the use of the alternative union syntax where possible.
+  Make sure to check the config options if you plan on using it!
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2.8.rst:28]: We should probably add the way to use it here (what to add in the pylintrc).
```suggestion
   (For example, 'typing.Dict' can be replaced by 'dict', and typing.Unions by ``|``, etc.)
  Make sure to check the config options if you plan on using it!
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I have some minor comments, but I think this is a great change. Let's put it in 2.8.0.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I was thinking of something like:
```python
@lru_cache()
def should_check_typing_alias(self):
    return self._py39_plus() or self._py37_plus() and self.config.runtime_typing is False
```

**Comment by cdce8p** [comment]
> I was thinking of something like:
> 
> ```python
> @lru_cache()
> def should_check_typing_alias(self):
>     return self._py39_plus() or self._py37_plus() and self.config.runtime_typing is False
> ```

Makes sense, I didn't thought of that one since we also need to check the node name after that. I think it's possible to combine it with my last refactoring.

**Comment by cdce8p** [comment]
With the latest change, it is quite elegant, I think.

--
If you don't have anything else, this MR would be ready. I would like to squash merge it if that's ok with you. A few mostly useless commit messages.

**Review by Pierre-Sassoulas — APPROVED** [review]
Yeap let's squash  :)

**Comment by cdce8p** [comment]
@KapJI FYI: Before merging this one, I added logic to detect if type aliases are used without subscript like you suggested here: https://github.com/home-assistant/core/pull/49480#discussion_r616796032

**Comment by KapJI** [comment]
Awesome, thanks! Happy to see these new features in pylint 🙂

**Pull request merged**